spotbugs/src/test/java/edu/umd/cs/findbugs/classfile/impl/ClassFactoryTest.java

```diff
@@ -0,0 +1,110 @@
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+
+import edu.umd.cs.findbugs.classfile.Global;
+import edu.umd.cs.findbugs.classfile.IScannableCodeBase;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ClassFactoryTest {
+
+    @Before
+    public void setUp () {
+        Global.setAnalysisCacheForCurrentThread(new NoopAnalysisCache());
+    }
+
+    @After
+    public void tearDown() {
+        Global.setAnalysisCacheForCurrentThread(null);
+    }
+
+    @Test
+    public void ignoreNonExistentFile() throws Exception {
+        File file = tempFile();
+        file.delete();
+        FilesystemCodeBaseLocator locator = buildLocator(file);
+        assertHasNoCodeBase(ClassFactory.createFilesystemCodeBase(locator));
+    }
+
+    @Test
+    public void ignoreUnreadableFile() throws Exception {
+        File file = createZipFile();
+        file.deleteOnExit();
+        file.setReadable(false);
+        FilesystemCodeBaseLocator locator = buildLocator(file);
+        assertHasNoCodeBase(ClassFactory.createFilesystemCodeBase(locator));
+    }
+
+    @Test
+    public void ignoreUnknownNonClassFileFormat() throws Exception {
+        FilesystemCodeBaseLocator locator = buildLocator(tempFile());
+        assertHasNoCodeBase(ClassFactory.createFilesystemCodeBase(locator));
+    }
+
+    @Test
+    public void acceptZipFile() throws Exception {
+        File zipFile = createZipFile();
+        zipFile.deleteOnExit();
+        FilesystemCodeBaseLocator locator = buildLocator(zipFile);
+        assertHasCodeBase(ClassFactory.createFilesystemCodeBase(locator));
+    }
+
+    private static File tempFile() throws Exception {
+        File tempFile = File.createTempFile("bug-497", ".txt");
+        tempFile.deleteOnExit();
+        return tempFile;
+    }
+
+    private static File createZipFile() throws Exception {
+        File zipFile = tempFile();
+        ZipOutputStream out = new ZipOutputStream(new FileOutputStream(zipFile));
+        ZipEntry entry = new ZipEntry("firstEntry");
+        out.putNextEntry(entry);
+        out.write("fileContents".getBytes(StandardCharsets.UTF_8));
+        out.closeEntry();
+        out.close();
+        return zipFile;
+    }
+
+    private FilesystemCodeBaseLocator buildLocator(File file) throws Exception {
+        return new FilesystemCodeBaseLocator(file.getCanonicalPath());
+    }
+
+    private void assertHasNoCodeBase(IScannableCodeBase codeBase) throws Exception {
+        assertNotNull(codeBase);
+        assertFalse(codeBase.iterator().hasNext());
+    }
+
+    private void assertHasCodeBase(IScannableCodeBase codeBase) throws Exception {
+        assertNotNull(codeBase);
+        assertTrue(codeBase.iterator().hasNext());
+    }
+}
```

Comment:
Which test case reproduces the bug reported at #497?

Reply:
Two of these could be represented by #497:  `ignoreUnreadableFile` and `ignoreUnknownNonClassFileFormat`. For the other two tests, an exception would be thrown which would generate a stacktrace in the output.